src/coreclr/vm/eventing/eventpipe/ep-rt-coreclr.h

```diff
@@ -2711,8 +2666,22 @@ ep_rt_diagnostics_command_line_get (void)
 {
 	STATIC_CONTRACT_NOTHROW;
 
-	diagnostics_command_line_lazy_init ();
-	return *diagnostics_command_line_get_ref ();
+	// In coreclr, this value can change over time, specifically before vs after suspension in diagnostics server.
+	// The host initalizes the runtime in two phases, init and exec assembly. On non-Windows platforms the commandline returned by the runtime
+	// is different during each phase. We suspend during init where the runtime has populated the commandline with a
+	// mock value (the full path of the executing assembly) and the actual value isn't populated till the exec assembly phase.
+	// On Windows this does not apply as the value is retrieved directly from the OS any time it is requested. 
+	// As a result, we cannot actually cache this value. We need to return the _current_ value.
+	// This function needs to handle freeing the string in order to make it consistent with Mono's version.
+	// To that end, we'll "cache" it here so we free the previous string when we get it again.
+	extern ep_char8_t *_ep_rt_coreclr_diagnostics_cmd_line;
+
+	if (_ep_rt_coreclr_diagnostics_cmd_line)
+		ep_rt_utf8_string_free(_ep_rt_coreclr_diagnostics_cmd_line);
```

Comment:
What guarantees that the caller is not using the string anymore when it is freed here?

Reply:
I think it would be better to make the caller responsible for freeing the string once it is no longer needed.

Reply:
Also, in shutdown you are not freeing this one anymore I think. You always seem to hold onto the latest one.

Reply:
This callstack looks like a potential use of the string that would occur concurrently with any usage from the diagnostic server thread:
ep_rt_diagnostics_command_line_get
log_process_info_event
disable_holding_lock
disable_helper
ep_disable
EP_RT_DEFINE_THREAD_FUNC (streaming_thread)

For practical purposes the odds of hitting that race are probably incredibly small, but at least theoretically I think the issue Jan was worried about can happen. It wouldn't make me too excited to introduce new potential AVs via servicing. If the value is known to only change once in the app lifetime I'd suggest it is safer to leak that one allocation rather than take risks freeing strings that could be in use.

Reply:
Thanks for spotting that, Noah! Feedback applied; please re-review 😄